transmission/transmission.go

```diff
@@ -84,57 +86,97 @@ func (h *Honeycomb) Start() error {
 	h.responses = make(chan Response, h.PendingWorkCapacity*2)
-	h.muster.MaxBatchSize = h.MaxBatchSize
-	h.muster.BatchTimeout = h.BatchTimeout
-	h.muster.MaxConcurrentBatches = h.MaxConcurrentBatches
-	h.muster.PendingWorkCapacity = h.PendingWorkCapacity
 	if h.Metrics == nil {
 		h.Metrics = &nullMetrics{}
 	}
-	h.muster.BatchMaker = func() muster.Batch {
-		return &batchAgg{
-			userAgentAddition: h.UserAgentAddition,
-			batches:           map[string][]*Event{},
-			httpClient: &http.Client{
-				Transport: h.Transport,
-				Timeout:   60 * time.Second,
-			},
-			blockOnResponse:       h.BlockOnResponse,
-			responses:             h.responses,
-			metrics:               h.Metrics,
-			disableCompression:    h.DisableGzipCompression || h.DisableCompression,
-			enableMsgpackEncoding: h.EnableMsgpackEncoding,
+	if h.batchMaker == nil {
+		h.batchMaker = func() muster.Batch {
+			return &batchAgg{
+				userAgentAddition: h.UserAgentAddition,
+				batches:           map[string][]*Event{},
+				httpClient: &http.Client{
+					Transport: h.Transport,
+					Timeout:   60 * time.Second,
+				},
+				blockOnResponse:       h.BlockOnResponse,
+				responses:             h.responses,
+				metrics:               h.Metrics,
+				disableCompression:    h.DisableGzipCompression || h.DisableCompression,
+				enableMsgpackEncoding: h.EnableMsgpackEncoding,
+			}
 		}
 	}
+
+	m := h.createMuster()
+	h.muster = m
 	return h.muster.Start()
 }
 
+func (h *Honeycomb) createMuster() *muster.Client {
+	m := new(muster.Client)
+	m.MaxBatchSize = h.MaxBatchSize
+	m.BatchTimeout = h.BatchTimeout
+	m.MaxConcurrentBatches = h.MaxConcurrentBatches
+	m.PendingWorkCapacity = h.PendingWorkCapacity
+	m.BatchMaker = h.batchMaker
+	return m
+}
+
 func (h *Honeycomb) Stop() error {
 	h.Logger.Printf("Honeycomb transmission stopping")
 	err := h.muster.Stop()
 	close(h.responses)
 	return err
 }
 
+func (h *Honeycomb) Flush() (err error) {
+	m := h.muster
+	defer func() {
+		e := m.Stop() // This is latent, so we do not want to execute this with the lock.
+		if err == nil {
+			err = e
+			// bubble up the error from Stop if there was no error from Start
+		}
+	}()
+	h.musterLock.Lock()
+	defer h.musterLock.Unlock()
+	newMuster := h.createMuster()
+	h.muster = newMuster
+	return h.muster.Start()
+}
+
 func (h *Honeycomb) Add(ev *Event) {
+	added := h.tryAdd(ev)
+	if added {
+		h.Metrics.Increment("messages_queued")
+		return
+	}
+	h.Metrics.Increment("queue_overflow")
+	r := Response{
+		Err:      errors.New("queue overflow"),
+		Metadata: ev.Metadata,
+	}
+	h.Logger.Printf("got response code %d, error %s, and body %s",
+		r.StatusCode, r.Err, string(r.Body))
+	writeToResponse(h.responses, r, h.BlockOnResponse)
+}
+
+// tryAdd attempts to add ev to the underlying muster. It returns false if this
+// was unsucessful because the muster queue (muster.Work) is full.
+func (h *Honeycomb) tryAdd(ev *Event) bool {
+	h.musterLock.RLock()
+	defer h.musterLock.RUnlock()
```

Comment:
There's an implication to holding the lock for the entirety of the `tryAdd` function here that I'd like called out explicitly in a comment somewhere. Maybe here to start but it might be something to push all the way up to user-facing documentation.

If you have `BlockOnSend` set to `true`, calling `Flush` will block until all pending `Add`s complete. 

I was wondering "Hmm, maybe we should take a local copy of the muster within the lock then release it"  aka:
```
h.musterLock.RLock()
localMuster := h.muster
h.musterLock.RUnlock()
...
...
localMuster.Work <- ev
```

This would avoid holding the lock while waiting to add events to a muster, and since the lock is primarily protecting _which_ running muster object is held by `h` it lets the `Add` continue to block on a muster even while also allowing the muster to be `Flush`ed. Except ... if this were the case, then a muster that got stopped while there was a pending goroutine trying to add an event to its `Work` channel might wind up sending existing events and closing that channel before the event got added - dropping the event. We promise not to let events leak here, so that seems like the wrong choice.

It feels uncomfortable to hold the lock for the entire time it takes to push an event into the channel when set to block on send, but I think the code here is correct and it must be done. I'd love another reviewer to also walk through this logic and confirm that it seems right.

Reply:
I added a comment to `Add` which notes the consequences of this lock. 

In case you were still thinking about the `localMuster` option, there is another issue here. This option will create a data race that could lead to a `panic`. Once the lock has been released, the `localMuster` could have its `Work` channel closed (through a concurrent call to `Stop` or `Flush`). If that happens, we'll `panic` when we try to send to the channel.

Reply:
I didn't say so in the comment above, but the fear of holding the lock for the entirety of `Add` is further reduced by it being a readlock so multiple `Add`s can be running simultaneously; it's only on `Flush` that there will be actual lock contention. The rest of the time there's additional bookkeeping, but it's the sort of bookkeeping that the `sync` package is built for, so nbd, right? It's not like `Add` is a super tight loop.